libraries/AP_WindVane/AP_WindVane.cpp

```diff
@@ -0,0 +1,320 @@
+    // @Description: set to one to enter calibration on reboot
+    // @Values: 0:None, 1:Calibrate
+    // @User: Standard
+    AP_GROUPINFO("CAL", 8, AP_WindVane, _calibration, 0),
+
+    // @Param: ANA_DZ
+    // @DisplayName: Analog potentiometer dead zone
+    // @Description: Analog potentiometer mechanical dead zone
+    // @Units: deg
+    // @Increment: 1
+    // @Range: 0 360
+    // @User: Standard
+    AP_GROUPINFO("ANA_DZ", 9, AP_WindVane, _analog_deadzone, 0),
+
+    AP_GROUPEND
+};
+
+// create a global instances of low pass filter
+LowPassFilterFloat low_pass_filter_wind_sin = LowPassFilterFloat(2.0f);
+LowPassFilterFloat low_pass_filter_wind_cos = LowPassFilterFloat(2.0f);
+LowPassFilterFloat low_pass_filter_wind_speed = LowPassFilterFloat(2.0f);
+
+// Public
+// ------
+
+// constructor
+AP_WindVane::AP_WindVane()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    if (_s_instance) {
+        AP_HAL::panic("Too many Wind Vane sensors");
+    }
+    _s_instance = this;
+}
+
+// destructor
+AP_WindVane::~AP_WindVane(void)
+{
+}
+
+/*
+ * Get the AP_WindVane singleton
+ */
+AP_WindVane *AP_WindVane::get_instance()
+{
+    return _s_instance;
+}
+
+// return true if wind vane is enabled
+bool AP_WindVane::enabled() const
+{
+    return (_type != WINDVANE_NONE);
+}
+
+// Initialize the Wind Vane object and prepare it for use
+void AP_WindVane::init()
+{
+    // a pin for reading the Wind Vane voltage
+    windvane_analog_source = hal.analogin->channel(ANALOG_INPUT_NONE);
+}
+
+// Update wind vane, called at 10hz
+void AP_WindVane::update()
+{
+    // exit immediately if not enabled
+    if (!enabled()) {
+        return;
+    }
+
+    // check for calibration
+    calibrate();
+
+    update_apparent_wind_direction();
+}
+
+// record home heading for use as wind direction if no sensor is fitted
+void AP_WindVane::record_home_headng()
+{
+    _home_heading = AP::ahrs().yaw;
+}
+
+// Private
+// -------
+
+// read the Wind Vane value from an analog pin, calculate bearing from analog voltage
+// assumes voltage increases as wind vane moves clockwise
+float AP_WindVane::read_analog()
+{
+    windvane_analog_source->set_pin(_analog_pin_no);
+    _current_analog_voltage = windvane_analog_source->voltage_average();
+
+    float current_analog_voltage_constrain = constrain_float(_current_analog_voltage,_analog_volt_min,_analog_volt_max);
+    float voltage_ratio = linear_interpolate(0.0f, 1.0f, current_analog_voltage_constrain, _analog_volt_min, _analog_volt_max);
+
+    float bearing = (voltage_ratio * radians(360 - _analog_deadzone)) + radians(_analog_head_bearing_offset);
+
+    return wrap_PI(bearing);
+}
+
+// read the bearing value from a PWM value on a RC channel (+- 45deg)
+float AP_WindVane::read_PWM_bearing()
+{
+    RC_Channel *ch = rc().channel(_rc_in_no-1);
+    if (ch == nullptr) {
+        return 0.0f;
+    }
+    float bearing = ch->norm_input() * radians(45);
```

Comment:
Is this really intended to always lock max input to +/- 45 degrees? Seems limiting/surprising (particularly since it appears this always feeds off current heading in the caller, which means you can't set a constant heading). Is this intended for some specific hardware rather then a RC transmitter providing the value?

Reply:
This is setup as a offset to the wind direction when armed, I initially tried it with +-180 but the resolution in a transmitter knob is so poor that it was impossible to set the angle correctly. I think +-45 is a good compromise between resolution and range, the wind is unlikely to shift by more than this.  The apparent wind does offset this by the current heading, this is used for setting the sails, the true wind direction for navigation uses this directly.

Reply:
I think the idea is that the pilot faces the vehicle into the wind at arming and this heading is recorded as the wind direction.  The transmitter input is just to allow slight adjustments to the heading.  I'm going to leave the limit at 45 degrees unless we get more input from users that they'd like the wider range.